File: recursion/capitalize_first.py
# Write a recursive function called capitalize_first. Given an array of strings, capitlize the first letter of each string in the array.
def capitalize_first(string_list):
    # Base case
    if len(string_list) == 0:
        return []
    # Recursion
    new_string = string_list[0].capitalize()
    return [new_string, *capitalize_first(string_list[1:])]


print(capitalize_first([]) == [])
print(capitalize_first(["car", "taco", "banana"]) == ["Car", "Taco", "Banana"])

File: recursion/capitalizeFirst.ts
// Write a recursive function called capitalizeFirst. Given an array of strings, capitlize the first letter of each string in the array.
import _ from 'lodash';

function capitalizeFirst(arr: string[]): string[] {
	// Base case
	if (arr.length === 0) return [];
	// Recursion
	const str: string = <string>arr[0];
	const newStr: string = (str[0] ?? '').toUpperCase() + str.slice(1);
	return [newStr, ...capitalizeFirst(arr.slice(1))];
}

console.log(_.isEqual(capitalizeFirst([]), []));
console.log(
	_.isEqual(capitalizeFirst(['car', 'taco', 'banana']), [
		'Car',
		'Taco',
		'Banana',
	])
);
